src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -64,23 +64,26 @@ public partial class JsonTypeInfo<T>
             CancellationToken cancellationToken,
             object? rootValueBoxed = null)
         {
-            return SerializeAsync(new AsyncSerializationStreamContext(utf8Json, Options), rootValue, cancellationToken, rootValueBoxed);
+            return SerializeAsync(new PooledByteBufferWriter(Options.DefaultBufferSize, utf8Json), rootValue, cancellationToken,
+                (int)(Options.DefaultBufferSize * JsonSerializer.FlushThreshold), rootValueBoxed);
         }
 
         internal Task SerializeAsync(PipeWriter utf8Json,
             T? rootValue,
             CancellationToken cancellationToken,
             object? rootValueBoxed = null)
         {
-            return SerializeAsync(new AsyncSerializationPipeContext(utf8Json), rootValue, cancellationToken, rootValueBoxed);
+            return SerializeAsync(utf8Json, rootValue, cancellationToken,
```

Comment:
Nit: could you extract the flushThreshold expression into its own line? (preferably with a comment explaining why it was chosen like that).